pkg/object/tcpproxy/connection.go

```diff
@@ -0,0 +1,395 @@
+	remoteAddr net.Addr
+
+	lastBytesSizeRead  int
+	lastWriteSizeWrite int
+
+	readBuffer      []byte
+	writeBuffers    net.Buffers
+	ioBuffers       []*iobufferpool.StreamBuffer
+	writeBufferChan chan *iobufferpool.StreamBuffer
+
+	mu               sync.Mutex
+	startOnce        sync.Once
+	connStopChan     chan struct{} // use for connection close
+	listenerStopChan chan struct{} // use for listener close
+
+	onRead  func(buffer *iobufferpool.StreamBuffer) // execute read filters
+	onClose func(event ConnectionEvent)
+}
+
+// NewClientConn wrap connection create from client
+func NewClientConn(conn net.Conn, listenerStopChan chan struct{}) *Connection {
+	return &Connection{
+		rawConn:          conn,
+		localAddr:        conn.LocalAddr(),
+		remoteAddr:       conn.RemoteAddr(),
+		listenerStopChan: listenerStopChan,
+
+		mu:              sync.Mutex{},
+		connStopChan:    make(chan struct{}),
+		writeBufferChan: make(chan *iobufferpool.StreamBuffer, writeBufSize),
+	}
+}
+
+// SetOnRead set connection read handle
+func (c *Connection) SetOnRead(onRead func(buffer *iobufferpool.StreamBuffer)) {
+	c.onRead = onRead
+}
+
+// SetOnClose set close callback
+func (c *Connection) SetOnClose(onclose func(event ConnectionEvent)) {
+	c.onClose = onclose
+}
+
+// Start running connection read/write loop
+func (c *Connection) Start() {
+	c.startOnce.Do(func() {
+		c.goWithRecover(func() {
+			c.startReadLoop()
+		}, func(r interface{}) {
+			_ = c.Close(NoFlush, LocalClose)
+		})
+
+		c.goWithRecover(func() {
+			c.startWriteLoop()
+		}, func(r interface{}) {
+			_ = c.Close(NoFlush, LocalClose)
+		})
+	})
+}
+
+func (c *Connection) goWithRecover(handler func(), recoverHandler func(r interface{})) {
+	go func() {
+		defer func() {
+			if r := recover(); r != nil {
+				logger.Errorf("tcp connection goroutine panic: %v\n%s\n", r, string(debug.Stack()))
+				if recoverHandler != nil {
+					// it is not needed to wrap recoverHandler with go func in the current scenario
+					recoverHandler(r)
+				}
+			}
+		}()
+		handler()
+	}()
+}
+
+// Write receive other connection data
+func (c *Connection) Write(buf *iobufferpool.StreamBuffer) (err error) {
+	defer func() {
+		if r := recover(); r != nil {
+			logger.Errorf("tcp connection has closed, local addr: %s, remote addr: %s, err: %+v",
+				c.localAddr.String(), c.remoteAddr.String(), r)
+			err = ErrConnectionHasClosed
+		}
+	}()
+
+	select {
+	case c.writeBufferChan <- buf:
+		return
+	default:
+	}
+
+	// try to send data again in 60 seconds
+	t := timerpool.Get(60 * time.Second)
+	select {
+	case c.writeBufferChan <- buf:
+	case <-t.C:
+		buf.Release()
+		err = ErrWriteBufferChanTimeout
+	}
+	timerpool.Put(t)
+	return
+}
+
+func (c *Connection) startReadLoop() {
+	for {
+		select {
+		case <-c.connStopChan:
+			return
+		case <-c.listenerStopChan:
+			return
+		default:
+			n, err := c.doReadIO()
+			if n > 0 {
+				c.onRead(iobufferpool.NewStreamBuffer(c.readBuffer[:n]))
+			}
+
+			if err != nil {
```

Comment:
early continue to reduce nesting.

```suggestion
			if err == nil {
			    continue
			}
```